src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,575 @@
+        )
+
+        return policy_loss_inputs, policy_opt_inputs
+
+    def _build_policy_loss(self, i):
+        policy_entropy = self._build_entropy_term(i)
+        rewards = i.reward_var
+
+        if self._maximum_entropy:
+            with tf.name_scope('augmented_rewards'):
+                rewards = i.reward_var + self.policy_ent_coeff * policy_entropy
+
+        with tf.name_scope('policy_loss'):
+            adv = compute_advantages(self.discount,
+                                     self.gae_lambda,
+                                     self.max_path_length,
+                                     i.baseline_var,
+                                     rewards,
+                                     name='adv')
+
+            adv_flat = flatten_batch_tensor(adv, name='adv_flat')
+            adv_valid = filter_valids_tensor(adv_flat,
+                                             i.flat.valid_var,
+                                             name='adv_valid')
+
+            if self.policy.recurrent:
+                adv = tf.reshape(adv, [-1, self.max_path_length])
+
+            # Optionally normalize advantages
+            eps = tf.constant(1e-8, dtype=tf.float32)
+            if self.center_adv:
+                if self.policy.recurrent:
+                    adv = center_advs(adv, axes=[0], eps=eps)
+                else:
+                    adv_valid = center_advs(adv_valid, axes=[0], eps=eps)
+
+            if self.positive_adv:
+                if self.policy.recurrent:
+                    adv = positive_advs(adv, eps)
+                else:
+                    adv_valid = positive_advs(adv_valid, eps)
+
+            # Calculate loss function and KL divergence
+            with tf.name_scope('kl'):
+                if self.policy.recurrent:
+                    kl = self.old_policy.distribution.kl_divergence(
+                        self.policy.distribution)
+                    pol_mean_kl = tf.reduce_sum(
+                        kl * i.valid_var) / tf.reduce_sum(i.valid_var)
+                else:
+                    kl = self.old_policy.distribution.kl_divergence(
+                        self.policy.distribution)
+                    pol_mean_kl = tf.reduce_mean(kl)
+
+            # Calculate vanilla loss
+            with tf.name_scope('vanilla_loss'):
+                if self.policy.recurrent:
+                    ll = self.policy.distribution.log_prob(
+                        i.action_var, name='log_likelihood')
+
+                    vanilla = ll * adv * i.valid_var
+                else:
+                    ll = self.policy.distribution.log_prob(
+                        i.valid.action_var, name='log_likelihood')
+
+                    vanilla = ll * adv_valid
+
+            # Calculate surrogate loss
+            with tf.name_scope('surrogate_loss'):
+                if self.policy.recurrent:
+                    lr = self.policy.distribution.prob(
```

Comment:
this is equivalent to 

```python
lr = tf.exp(self.policy.distribution.log_prob(i.action_var) - tf.stop_gradient(self.old_policy.distribution.log_prob(i.action_var)))
```
and we generally use that form for numerical reasons.